File: src/LegacyVault.Web/src/crypto/keyCache.ts
/**
 * In-memory session key cache.
 * Keys are never persisted to localStorage/sessionStorage/IndexedDB.
 * Cache is automatically cleared when the page unloads.
 */

const cache = new Map<string, CryptoKey>();

export const keyCache = {
  set: (categoryId: string, key: CryptoKey) => cache.set(categoryId, key),
  get: (categoryId: string) => cache.get(categoryId),
  has: (categoryId: string) => cache.has(categoryId),
  clear: (categoryId: string) => cache.delete(categoryId),
  clearAll: () => cache.clear(),
};

// Clear all keys on page unload for security
window.addEventListener('beforeunload', () => cache.clear());

File: src/LegacyVault.iOS/lib/core/crypto/key_cache.dart
import 'dart:typed_data';

/// In-memory cache of derived encryption keys, keyed by category ID.
/// Keys are never persisted to disk.
class KeyCache {
  KeyCache._();
  static final KeyCache instance = KeyCache._();

  final Map<String, Uint8List> _cache = {};

  void store(String categoryId, Uint8List key) {
    _cache[categoryId] = key;
  }

  Uint8List? get(String categoryId) => _cache[categoryId];

  bool has(String categoryId) => _cache.containsKey(categoryId);

  void remove(String categoryId) => _cache.remove(categoryId);

  void clearAll() => _cache.clear();
}
